src/main/java/graphql/annotations/GraphQLAnnotations.java

```diff
@@ -495,24 +496,35 @@ private boolean checkIfPrefixGetterExists(Class c, String prefix, String propert
         return true;
     }
 
-    private GraphQLOutputType getGraphQLConnection(boolean isConnection, AccessibleObject field, GraphQLOutputType type, GraphQLOutputType outputType, GraphQLFieldDefinition.Builder builder) {
-        if (isConnection) {
-            if (type instanceof GraphQLList) {
-                graphql.schema.GraphQLType wrappedType = ((GraphQLList) type).getWrappedType();
-                assert wrappedType instanceof GraphQLObjectType;
-                String annValue = field.getAnnotation(GraphQLConnection.class).name();
-                String connectionName = annValue.isEmpty() ? wrappedType.getName() : annValue;
-                GraphQLObjectType edgeType = RELAY_TYPES.edgeType(connectionName, (GraphQLOutputType) wrappedType, null, Collections.<GraphQLFieldDefinition>emptyList());
-                outputType = RELAY_TYPES.connectionType(connectionName, edgeType, Collections.emptyList());
-                builder.argument(RELAY_TYPES.getConnectionFieldArguments());
-            }
+    private GraphQLOutputType getGraphQLConnection(AccessibleObject field, GraphQLOutputType type, GraphQLFieldDefinition.Builder builder) {
+        boolean isNonNull = (type instanceof GraphQLNonNull);
+        if (isNonNull) {
+            type = (GraphQLOutputType) ((GraphQLNonNull) type).getWrappedType();
+        }
+
+        if (type instanceof GraphQLList) {
+            graphql.schema.GraphQLType wrappedType = ((GraphQLList) type).getWrappedType();
+            assert wrappedType instanceof GraphQLObjectType;
+            String annValue = field.getAnnotation(GraphQLConnection.class).name();
+            String connectionName = annValue.isEmpty() ? wrappedType.getName() : annValue;
+            GraphQLObjectType edgeType = RELAY_TYPES.edgeType(connectionName, (GraphQLOutputType) wrappedType, null, Collections.<GraphQLFieldDefinition>emptyList());
+            type = RELAY_TYPES.connectionType(connectionName, edgeType, Collections.emptyList());
+            builder.argument(RELAY_TYPES.getConnectionFieldArguments());
         }
-        return outputType;
+
+        if (isNonNull) {
+            type = new GraphQLNonNull(type);
+        }
+        return type;
     }
 
-    private boolean isConnection(AccessibleObject obj, Class<?> klass, GraphQLOutputType type) {
+    private boolean isConnection(AccessibleObject obj, GraphQLOutputType type) {
+        if (type instanceof GraphQLNonNull) {
+            type = (GraphQLOutputType) ((GraphQLNonNull) type).getWrappedType();
+        }
+        final GraphQLOutputType wrappedType = type;
```

Comment:
Can you rename wrappedType to actualType? It is not really a wrapped type (it is probably a list)
I dont like the line ``aClass -> aClass.isInstance(((GraphQLList) wrappedType).getWrappedType())``

Reply:
Do you prefer : 
```
        if (obj.isAnnotationPresent(GraphQLConnection.class) && actualType instanceof GraphQLList) {
            graphql.schema.GraphQLType wrappedType = ((GraphQLList) actualType).getWrappedType();
            return TYPES_FOR_CONNECTION.stream().anyMatch(aClass -> aClass.isInstance(wrappedType));
        }
        return false;
```
It's a little more verbose , buy maybe better ?